src/gc-arena/src/cell.rs

```diff
@@ -0,0 +1,152 @@
+use core::{
+    cell::{BorrowError, BorrowMutError, Cell, Ref, RefCell, RefMut},
+    fmt::{self, Debug},
+};
+
+use crate::{Collect, CollectionContext};
+
+/// A `Cell` type that is `Collect`.
+///
+/// This does not trigger a write barrier on writes, so the only methods that allow writing to the
+/// inner type are unsafe.
+///
+/// SAFETY: In order to maintain the invariants of the garbage collector, no new `Gc` pointers
+/// may be adopted by this type as a result of any method that affords interior mutability, unless
+/// the write barrier for the containing `Gc` pointer is invoked manually before collection is
+/// triggered.
+pub struct CollectCell<T: ?Sized> {
+    cell: Cell<T>,
+}
+
+impl<T: Copy + Debug> fmt::Debug for CollectCell<T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_tuple("CollectCell").field(&self.cell).finish()
+    }
+}
+
+impl<T: Copy> Clone for CollectCell<T> {
+    fn clone(&self) -> Self {
+        Self {
+            cell: self.cell.clone(),
+        }
+    }
+}
+
+impl<T: ?Sized> CollectCell<T> {
+    pub fn as_ptr(&self) -> *mut T {
+        self.cell.as_ptr()
+    }
+}
+
+impl<T> CollectCell<T> {
+    pub fn new(t: T) -> CollectCell<T> {
+        Self { cell: Cell::new(t) }
+    }
+
+    /// Call `Cell::set` on the inner `Cell`.
+    ///
+    /// SAFETY: See [`CollectCell`]
+    pub unsafe fn set(&self, t: T) {
+        self.cell.set(t);
+    }
+
+    /// Call `Cell::replace` on the inner `Cell`.
+    ///
+    /// SAFETY: See [`CollectCell`]
+    pub unsafe fn replace(&self, t: T) -> T {
+        self.cell.replace(t)
+    }
+}
+
+impl<T: Copy> CollectCell<T> {
+    pub fn get(&self) -> T {
+        self.cell.get()
+    }
+}
+
+unsafe impl<'gc, T: Collect + Copy + 'gc> Collect for CollectCell<T> {
+    fn trace(&self, cc: CollectionContext) {
+        self.cell.get().trace(cc);
+    }
+}
```

Comment:
This is a very degenerate case, but this is incorrect if `T::trace` modifies the inner `T`; the potentially modified value should be written back to the `Cell`:
```rs
fn trace(&self, cc: CollectionContext) {
    // Use a guard that writes back the value in case the 'trace' call mutates it.
    let cell = &self.cell;
    let guard = scopeguard::guard(cell.get(), |copied| cell.set(copied));
    T::trace(&guard, cc);
}
```